hack/cloud-image-uploader/azure.go

```diff
@@ -0,0 +1,420 @@
+	resourceGroupName = "SideroGallery"
+	defaultRegion = "eastus"
+	galleryName = "SideroLabs"
+	storageAccount = "siderogallery"
+)
+
+type TargetRegion struct {
+	Name                 string `json:"name"`
+	RegionalReplicaCount int    `json:"regionalReplicaCount"`
+	StorageAccountType   string `json:"storageAccountType"`
+}
+
+// Mapping CPU architectures to Azure architectures
+var azureArchitectures = map[string]string{
+	"amd64": "x64",
+	"arm64": "arm64",
+}
+
+// AzureUploader represents an object that has the capability to upload to Azure
+type AzureUploader struct {
+	Options Options
+}
+
+func (azu *AzureUploader) AzureGalleryUpload(ctx context.Context) error {
+	var err error
+
+	var g *errgroup.Group
+	g, ctx = errgroup.WithContext(ctx)
+
+	fmt.Println("Setting default creds")
+
+	err = setDefaultAzureCreds()
+	if err != nil {
+		log.Fatalf("Error: %v\n", err)
+	}
+
+	fmt.Println("Getting locations")
+
+	err = getAzureLocations(ctx)
+	if err != nil {
+		log.Fatalf("Error: %v\n", err)
+	}
+
+	// Upload blob
+	fmt.Printf("Architectures: %+v\n", azu.Options.Architectures)
+
+	for _, arch := range azu.Options.Architectures {
+		arch := arch
+
+		g.Go(func() error {
+			fmt.Printf("Starting upload blob for %s\n", arch)
+			err = azu.uploadAzureBlob(ctx, arch)
+			if err != nil {
+				return fmt.Errorf("error uploading page blob for %s: %w", arch, err)
+			}
+			fmt.Printf("Starting disk creation for %s\n", arch)
+			err = azu.createAzureDisk(ctx, "azure-disk-template.json", arch)
+			if err != nil {
+				fmt.Printf("Error: %v\n", err)
+			}
+			fmt.Printf("Starting image version creation for %s\n", arch)
+			err = azu.createAzureImageVersion(ctx, "azure-image-version-template.json", arch)
+			if err != nil {
+				fmt.Printf("Error: %v\n", err)
+			}
+			return err
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		fmt.Println(err)
+	}
+
+	return nil
+}
+
+func (azu *AzureUploader) uploadAzureBlob(ctx context.Context, arch string) error {
+
+	var g *errgroup.Group
+	g, ctx = errgroup.WithContext(ctx)
+
+	blobURL := fmt.Sprintf("https://siderogallery.blob.core.windows.net/images/talos/talos-%s-%s.vhd", arch, azu.Options.Tag)
+
+	pageBlobClient, err := pageblob.NewClient(blobURL, cred, nil)
+	if err != nil {
+		log.Printf("Error: %v\n", err)
+	}
+
+	source, err := os.Open(azu.Options.AzureImage(arch))
+	if err != nil {
+		return err
+	}
+
+	defer func(source *os.File) {
+		err := source.Close()
+		if err != nil {
+
+		}
+	}(source) //nolint:errcheck
+
+	file, err := ExtractFileFromTarGz("disk.vhd", source)
+	if err != nil {
+		return err
+	}
+
+	data, err := io.ReadAll(file)
+	if err != nil {
+		log.Printf("Error: %v\n", err)
+		return err
+	}
+
+	reader := bytes.NewReader(data)
+	totalSize := int64(reader.Len())
+
+	// Check if the file size is a multiple of 512 bytes
+	if totalSize%pageblob.PageBytes != 0 {
+		panic("The file size must be a multiple of 512 bytes")
+	}
+
+	_, err = pageBlobClient.Create(ctx, totalSize, nil)
```

Comment:
this whole upload function with chunking seems to be really complicated, and this is all already implemented in the SDK

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/storage/azblob#example-package-Client_UploadStream

they have options for concurrency, page size, etc. do we really need to re-implement it?

Reply:
The link you shared uses the UploadStream function which will upload the image file as a Block Blob. The file must be uploaded as a Page Blob to used to create managed disks and images. There does not seem to be anything similar to UploadStream in the SDK for Page Blobs.

Reply:
ok, makes sense, in that case the code needs to be rewritten, and simplieid, as it's turned inside out:

* number of goroutines should be number of concurrent uploaders
* channel should contain "work" items - [start_offset, end_offset)
* each worker picks up the work from the channel, uploads and looks for the next check

There's no point to create 1000 goroutines and make them all wait on a semaphore.